### ISSUE TITLE
How does dowsample_fraction work?

### ISSUE THREAD

**Issue by anands-repo** [google/deepvariant#230]
I looked at https://github.com/google/deepvariant/blob/r0.8/docs/deepvariant-details-training-data.md, where it says "and, more `dowsample_fraction` during training". Looking through make_examples it seems `downsample_fraction` is a fixed number for any given run of make_examples. My assumption currently is that make_examples is run multiple times with different `downsample_fraction` settings to obtain a wider mix of coverages in the training data, before running training (please correct me if wrong).

I also had questions of how the downsampled mix was selected. E.g., were coverages in [20, 60]x equally distributed, or were the lower coverages used sparingly (since, for example, 20x is not a very interesting case)? Also, training a single model on multiple coverages and testing (variant calling) on a single coverage could be problematic (but in DeepVariant it is not), because the data distribution is different between the training and test sets.

If there is a blog or article with some of these details, that would also be great!

**Comment by sidharthgoel** [comment]
On the first point, your understanding is correct. We have separate runs with the fraction set to different values, and these examples are merged and shuffled. Also see [this description](https://github.com/google/deepvariant/blob/aff131aac3bd0cb63ee8314e32bcbf5590987fb8/deepvariant/make_examples.py#L163) from make_examples:

> Reads will be kept (randomly) with a probability of downsample_fraction from the input BAM. This argument makes it easy to create examples as though the input BAM had less coverage.

For the second point, @AndrewCarroll can share some details.

**Comment by AndrewCarroll** [comment]
Hi @anands-repo 

With respect to your second point - The starting training data begins from various coverages, ranging roughly from 27x-60x. From the starting coverage of the BAM files, downsampling is applied in 0.1 increments until the coverage would reach around 20x. As a result, higher coverage ranges are slightly less represented than the 30x-20x range, but not by a substantial amount.

There are two main purposes of this - the first is to allow DeepVariant to perform well across many different coverages. The second is to ensure that there are many hard examples to learn from. Our strategy in downsampling is not fixed - for example, as we add more training data in the future, we may decide to have fewer downsample increments to generally keep the same range of examples. The coverage range has also evolved over time, for example in v0.7, the range was roughly 60x-30x. Extending the range to 20x caused a very small decline in accuracy at 50x (and also therefore the case study), but resulted in more substantial gains in the 20x-30x range.

I would stress that there isn't one correct downsampling strategy, as long as it represents a diversity of coverages and helps create hard examples.

**Comment by anands-repo** [comment]
Thanks @sidharthgoel @AndrewCarroll ! Just one more question ...

Current subsampling is static in the sense, the training loops see the same images every epoch. Just wanted to ask, did you consider doing a dynamic type of random subsampling, such that reads are selected just before entering the DNN from a fixed make_examples dump?

**Comment by AndrewCarroll** [comment]
Hi @anands-repo 

You are correct, subsampling is static. A nice effect of this is that it reduces the complexity in terms of training reproducibility. We have not deeply investigated whether dynamic resampling (making a different image per epoch) would benefit training. It's an interesting question. It could potentially reduce overfitting that might occur at the read level and therefore allow training to progress through more epochs before a model is selected. 

I think it is unlikely that this would improve the current production training setup, but it is not impossible. For the WGS training curves, there is little overfitting apparent in training graphs over a large number of epochs. For the WES training curves, some overfitting is apparent, but we suspect this is less due to signal from the read level and more due to the smaller number of regions represented in the exome. This is one reason that we currently train the exome model by warmstarting from the WGS model.

It is probably worth us taking a look at some point, but likely isn't the lowest hanging fruit for us to improve performance.

Thank you for the suggestion and discussion.

**Comment by anands-repo** [comment]
@AndrewCarroll Thanks for sharing your thoughts. I understand regarding the complexities involved, and that this may not be a low-hanging fruit. Also, bigger gains may be had from better data I imagine. While GIAB and Platinum Genomes are great, it would be great if the Syndip dataset would be pulled into a similar effort such as GIAB.

Regarding dynamic subsampling, there may also be a need to correct for some effects such as re-thresholding with the 0.12 allele fraction cut-off after subsampling etc. The effect of realignment at a lower coverage (coverage after subsampling) may come into picture as well though may be not that much. Dynamic subsampling could result in a different input data statistic overall. May be your original assessment that static subsampling reduces complexity is a good enough argument to just continue with it.

Thanks for the discussion!
